java/src/org/openqa/selenium/grid/node/config/NodeOptions.java

```diff
@@ -344,143 +344,124 @@ private SessionFactory createSessionFactory(String clazz, Capabilities stereotyp
-                      .toArray();
-
-              if (configIndexes.length == 0) {
-                throw new ConfigException(
-                    "No 'display-name' keyword was found in the provided configs!");
-              }
-
-              List<Map<String, String>> driversMap = new ArrayList<>();
-              for (int i = 0; i < configIndexes.length; i++) {
-                int fromIndex = configIndexes[i];
-                int toIndex =
-                    (i + 1) >= configIndexes.length ? drivers.size() : configIndexes[i + 1];
-                Map<String, String> configMap = new HashMap<>();
-                drivers
-                    .subList(fromIndex, toIndex)
-                    .forEach(
-                        keyValue -> {
-                          String[] values = keyValue.split("=", 2);
-                          configMap.put(values[0], unquote(values[1]));
-                        });
-                driversMap.add(configMap);
-              }
+    Multimap<WebDriverInfo, SessionFactory> driverConfigs = HashMultimap.create();
 
-              List<DriverService.Builder<?, ?>> builders = new ArrayList<>();
-              ServiceLoader.load(DriverService.Builder.class).forEach(builders::add);
+    // get all driver configuration settings
+    config.getAll(NODE_SECTION, "driver-configuration")
+      // if settings exist
+      .ifPresent(drivers -> {
+        Map<String, String> configMap = new HashMap<>();
+        List<Map<String, String>> configList = new ArrayList<>();
+
+        // iterate over driver settings
+        for (String setting : drivers) {
+          // split this setting into key/value pair
+          String[] values = setting.split("=", 2);
+          // if format is invalid
+          if (values.length != 2) {
+            throw new ConfigException("Driver setting '" + setting
+              + "' does not adhere to the required 'key=value' format!");
+          }
+          // if this is a record separator
+          if (values[0].equals(Config.DELIM_KEY)) {
+            // if config lacks settings
+            if (configMap.isEmpty()) {
+              throw new ConfigException("Found config delimiter with no preceding settings!");
+            }
+
+            // if config lacks 'display-name' setting
+            if (!configMap.containsKey("display-name")) {
+              throw new ConfigException("Found config with no 'display-name' setting! " + configMap);
+            }
+
+            // if config lacks 'stereotype' setting
+            if (!configMap.containsKey("stereotype")) {
+              throw new ConfigException("Found config with no 'stereotype' setting! " + configMap);
+            }
+
+            // add config to list
+            configList.add(configMap);
+            // prepare for next config
+            configMap = new HashMap<>();
+          } else {
+            // add setting to config
+            configMap.put(values[0], unquote(values[1]));
```

Comment:
It would be quite easy to ensure that only supported key names are used, and that each key is only specified once.